README.rst

TargetProcess Client

Python library to help getting data from `TargetProcess API <http://dev.targetprocess.com/rest/getting_started>`_

.. image:: https://img.shields.io/pypi/v/targetprocess-client.svg
    :target: https://badge.fury.io/py/targetprocess-client
    :alt: Pypi

.. image:: https://travis-ci.org/360youlun/targetprocess-client.svg?branch=master
    :target: https://travis-ci.org/360youlun/targetprocess-client
    :alt: Travis CI

.. image:: https://codecov.io/github/360youlun/targetprocess-client/coverage.svg?branch=master
    :target: https://codecov.io/github/360youlun/targetprocess-client?branch=master
    :alt: Codecov

.. image:: https://www.quantifiedcode.com/api/v1/project/8cdc9e5652dd4aaf8c8465b788966ea3/badge.svg
    :target: https://www.quantifiedcode.com/app/project/8cdc9e5652dd4aaf8c8465b788966ea3
    :alt: Code issues

================
Getting the code
================

The code is hosted at https://github.com/magicjohnson/targetprocess-client

Check out the latest development version with::

    $ git clone git://github.com/magicjohnson/targetprocess-client.git

==========
Installing
==========

You can install targetprocess-client using::

    $ pip install targetprocess-client

or get the code and run install::

    $ cd targetprocess-client
    $ python setup.py install

==============
Usage examples
==============

Create client instance:

.. code-block:: python

    from targetprocess.api import TargetProcessAPIClient
    from targetprocess.serializers import TargetProcessSerializer
    tp = TargetProcessAPIClient(api_url='https://md5.tpondemand.com/api/v1/', user='user', password='pass')

Get collection of UserStories:

.. code-block:: python

    stories_json = tp.get_stories(take=5, include="[Id,Name,CreateDate]")
    stories = TargetProcessSerializer.deserialize(stories_json)

Get UserStory item

.. code-block:: python

    story_json = tp.get_story(360, include="[Id,Name,CreateDate]")
    story = TargetProcessSerializer.deserialize(story_json)

Update item

.. code-block:: python

    data = {'Name': 'New name'}
    tp.update_story(360, data)

Request "unregistered" collection (that client doesn't have predefined methods for)

.. code-block:: python

    tp.get_collection(collection="Epics", take=2)
    tp.get_resource(collection="Epics", entity_id=427)
    tp.update_resource(collection="Epics", entity_id=427, data={'Name': 'New Epic name'})


CHANGELOG.md
# Change Log

## [v0.1.2](https://github.com/magicjohnson/targetprocess-client/tree/v0.1.2) (2016-04-07)
[Full Changelog](https://github.com/magicjohnson/targetprocess-client/compare/v0.1.1...v0.1.2)

**Merged pull requests:**

- Don't raise error in case of 201 response [\#4](https://github.com/magicjohnson/targetprocess-client/pull/4) ([magicjohnson](https://github.com/magicjohnson))

## [v0.1.1](https://github.com/magicjohnson/targetprocess-client/tree/v0.1.1) (2016-03-12)
[Full Changelog](https://github.com/magicjohnson/targetprocess-client/compare/v0.1.0...v0.1.1)

**Merged pull requests:**

- python3 support [\#2](https://github.com/magicjohnson/targetprocess-client/pull/2) ([magicjohnson](https://github.com/magicjohnson))
- README update [\#1](https://github.com/magicjohnson/targetprocess-client/pull/1) ([magicjohnson](https://github.com/magicjohnson))

## [v0.1.0](https://github.com/magicjohnson/targetprocess-client/tree/v0.1.0) (2016-03-12)


\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

tests/tests_api.py
# coding=utf-8
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

from unittest import TestCase

import mock
from requests import ConnectionError

from targetprocess.api import TargetProcessAPIClient
from targetprocess.exceptions import BadResponseError


class TargetProcessAPIClientTest(TestCase):
    def setUp(self):
        self.maxDiff = None
        self.api = TargetProcessAPIClient('http://tp.api.url/api/v1', 'username', 'password')
        self.do_request_patcher = mock.patch.object(self.api, '_do_request', autospec=True)
        self.do_request_mock = self.do_request_patcher.start()
        self.addCleanup(self.do_request_patcher.stop)

        self.response = {'Items': [{'Id': 1}, {'Id': 2}]}
        self.do_request_mock.return_value = self.response

    def test_get_story_returns_story_by_id(self):
        self.assertEqual(self.api.get_story(8321), self.response)
        self.do_request_mock.assert_called_once_with(
            method='get',
            url="http://tp.api.url/api/v1/UserStories/8321?format=json"
        )

    def test_get_stories(self):
        response = self.api.get_stories(where="Team.Name eq 'Team X'")
        self.assertEqual(response, self.response)
        self.do_request_mock.assert_called_once_with(
            method='get',
            url="http://tp.api.url/api/v1/UserStories/?take=20&where=Team.Name+eq+%27Team+X%27&format=json",
        )

    def test_pagination(self):
        self.do_request_mock.side_effect = [
            {
                'Items': [{'Id': 3}, {'Id': 4}],
                'Next': 'http://tp.api.url/api/v1/UserStories/?take=20&skip=20&format=json'
            },
            self.response,
        ]

        response = self.api.get_stories()
        self.assertEqual(response['Items'], [{'Id': 3}, {'Id': 4}, {'Id': 1}, {'Id': 2}])
        self.assertSequenceEqual(
            self.do_request_mock.mock_calls, [
                mock.call(method='get', url="http://tp.api.url/api/v1/UserStories/?take=20&format=json"),
                mock.call(method='get', url="http://tp.api.url/api/v1/UserStories/?take=20&skip=20&format=json"),
            ]
        )

    def test_nested_pagination(self):
        response_with_next = {
            'Items': [
                {
                    'Id': 1,
                    'Nested': {
                        'Items': [{'Id': 3}],
                        'Next': 'http://tp.api.url/api/v1/UserStories/1/Nested/?take=25&skip=25',
                    }
                },
                {
                    'Id': 2,
                    'Nested': {
                        'Items': [{'Id': 3}],
                        'Next': 'http://tp.api.url/api/v1/UserStories/2/Nested/?take=25&skip=25',
                    }
                }
            ]
        }
        self.do_request_mock.side_effect = [
            response_with_next,
            self.response,
            self.response,
        ]

        response = self.api.get_stories()
        self.assertEqual(response, {
            'Items': [
                {
                    'Id': 1,
                    'Nested': {
                        'Items': [{'Id': 3}, {'Id': 1}, {'Id': 2}],
                        'Next': 'http://tp.api.url/api/v1/UserStories/1/Nested/?take=25&skip=25'
                    },
                },
                {
                    'Id': 2,
                    'Nested': {
                        'Items': [{'Id': 3}, {'Id': 1}, {'Id': 2}],
                        'Next': 'http://tp.api.url/api/v1/UserStories/2/Nested/?take=25&skip=25',
                    }
                },
            ]
        })
        expected_calls = [
            mock.call(method='get', url="http://tp.api.url/api/v1/UserStories/?take=20&format=json"),
            mock.call(method='get', url="http://tp.api.url/api/v1/UserStories/1/Nested/?take=25&skip=25&format=json"),
            mock.call(method='get', url="http://tp.api.url/api/v1/UserStories/2/Nested/?take=25&skip=25&format=json"),
        ]
        self.assertSequenceEqual(self.do_request_mock.mock_calls, expected_calls)

    def test_update_story(self):
        self.assertEqual(self.api.update_story(1234, {'data': 'xxx'}), self.response)
        self.do_request_mock.assert_called_once_with(
            method='post',
            url='http://tp.api.url/api/v1/UserStories/1234?include=%5BId%5D&format=json',
            data='{"data": "xxx"}',
            headers={'Content-type': 'application/json'},
        )


class APIClientDoRequestTest(TestCase):
    def setUp(self):
        self.requests_patcher = mock.patch('targetprocess.api.requests')
        self.requests_mock = self.requests_patcher.start()
        self.addCleanup(self.requests_patcher.stop)

        getattr = self.requests_mock.__getattribute__ = mock.MagicMock()
        self.method = getattr.return_value
        self.response = self.method.return_value
        self.response.status_code = 200
        self.response.json.return_value = {'Items': [{'Id': 1}, {'Id': 2}, {'Id': 3}]}
        self.api = TargetProcessAPIClient('http://tp.api.url/api/v1', 'username', 'password')

    def test_request_retry(self):
        self.method.side_effect = ConnectionError
        self.assertEqual(self.api.get_stories(), {})
        self.assertEqual(self.method.mock_calls, [
            mock.call(url='http://tp.api.url/api/v1/UserStories/?take=20&format=json', auth=self.api.auth),
            mock.call(url='http://tp.api.url/api/v1/UserStories/?take=20&format=json', auth=self.api.auth),
        ])

    def test_do_request_returns_response(self):
        response = self.api.get_stories()
        self.assertEqual(response, {'Items': [{'Id': 1}, {'Id': 2}, {'Id': 3}]})
        self.method.assert_called_once_with(
            url='http://tp.api.url/api/v1/UserStories/?take=20&format=json', auth=self.api.auth
        )

    def test_do_request_raises_bad_response(self):
        self.response.status_code = 400
        with self.assertRaises(BadResponseError):
            self.api.get_stories()

        self.method.assert_called_once_with(
            url='http://tp.api.url/api/v1/UserStories/?take=20&format=json', auth=self.api.auth
        )


tests/tests_serializers.py
# coding=utf-8
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

from datetime import datetime
from unittest import TestCase

from pytz import UTC

from targetprocess.serializers import TargetProcessSerializer


class TargetProcessSerializerTest(TestCase):
    def setUp(self):
        self.maxDiff = None

    def test_deserialize_dict(self):
        data = {
            'EndDate': '/Date(1441596445000-0500)/',
            'Effort': 0.0,
            'ResourceType': 'UserStory',
            'Team': {
                'Id': 298,
                'Name': 'DevOps',
            },
            'LastCommentDate': None,
            'CustomFields': [
                {
                    'Name': 'UI Spec',
                    'Type': 'RichText',
                    'Value': None
                },
                {
                    'Name': 'Date',
                    'Type': 'DropDown',
                    'Value': '/Date(1441596445000-0500)/'
                },
            ]
        }

        expected = {
            'EndDate': datetime(2015, 9, 7, 3, 27, 25, tzinfo=UTC),
            'Effort': 0.0,
            'ResourceType': 'UserStory',
            'Team': {
                'Id': 298,
                'Name': 'DevOps',
            },
            'LastCommentDate': None,
            'CustomFields': [
                {
                    'Name': 'UI Spec',
                    'Type': 'RichText',
                    'Value': None
                },
                {
                    'Name': 'Date',
                    'Type': 'DropDown',
                    'Value': datetime(2015, 9, 7, 3, 27, 25, tzinfo=UTC)
                },
            ]
        }
        result = TargetProcessSerializer().deserialize(data)
        self.assertEqual(result, expected)

    def test_deserialize_dict_with_items(self):
        data = {
            'Items': [
                {
                    'Date': '/Date(1441596445000-0500)/',
                    'NoneField': None,
                    'TextField': 'Text',
                    'NestedDict': {'Field': 'Value'},
                    'NestedList': [{'Field': 'Value'}]
                }
            ]
        }
        expected = [
            {
                'Date': datetime(2015, 9, 7, 3, 27, 25, tzinfo=UTC),
                'NoneField': None,
                'TextField': 'Text',
                'NestedDict': {'Field': 'Value'},
                'NestedList': [{'Field': 'Value'}]
            }
        ]

        result = TargetProcessSerializer().deserialize(data)
        self.assertEqual(result, expected)


targetprocess/serializers.py
# coding=utf-8
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

from datetime import datetime

import pytz
from six import string_types


class TargetProcessSerializer(object):
    """
    de-serializes raw data.
    For collection removes redundant `Items` key and puts everything into root
    converts weird "json" timestamp format into python datetime object
    """
    @classmethod
    def deserialize(cls, data):
        if isinstance(data, list):
            return [cls.deserialize(item) for item in data]

        if isinstance(data, dict):
            if 'Items' in data:
                return [cls.deserialize(item) for item in data['Items']]

            processed_item = {}
            for key, value in data.items():
                if isinstance(value, (dict, list)):
                    processed_item[key] = cls.deserialize(value)
                elif cls._is_date(value):
                    processed_item[key] = cls._json_date_to_datetime(value)
                else:
                    processed_item[key] = value

            return processed_item

        return data

    @staticmethod
    def _is_date(value):
        return isinstance(value, string_types) and value.startswith('/Date')

    @staticmethod
    def _json_date_to_datetime(json_date):
        """
        Returns timestamp for a JSON date format
        """
        sign = json_date[-7]
        if sign not in '-+' or len(json_date) == 13:
            milli_secs = int(json_date[6:-2])
        else:
            milli_secs = int(json_date[6:-7])

        ts = milli_secs / 1000

        d = datetime.fromtimestamp(ts, tz=pytz.UTC)
        return d


targetprocess/api.py
# coding=utf-8

"""
TargetProcess API Client
"""
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

import json
import logging
import sys
from collections import OrderedDict

if sys.version_info[0] < 3:
    from urllib import urlencode
    from urlparse import parse_qsl, urlparse, urlunparse
else:
    from urllib.parse import urlencode
    from urllib.parse import parse_qsl, urlparse, urlunparse


import requests

from requests.auth import HTTPBasicAuth
from requests.exceptions import ConnectionError, ReadTimeout

from six import add_metaclass

from targetprocess.exceptions import BadResponseError

logger = logging.getLogger(__name__)


class MetaAPI(type):
    """
    Meta class that magically registers some commonly (at least for me) used collections
    by adding methods to TargetProcessAPIClient class
    """
    RESOURCES = (
        ('story', 'UserStories'),
        ('bug', 'Bugs'),
        ('release', 'Releases'),
        ('role_effort', 'RoleEfforts'),
        ('user', 'Users'),
    )
    COLLECTIONS = (
        ('stories', 'UserStories'),
        ('bugs', 'Bugs'),
        ('releases', 'Releases'),
        ('iterations', 'TeamIterations'),
        ('members', 'TeamMembers'),
    )

    def __init__(cls, name, bases, dct):

        cls._add_resource_methods()
        cls._add_collection_methods()

        super(MetaAPI, cls).__init__(name, bases, dct)

    def _add_resource_methods(cls):
        for resource, tp_name in cls.RESOURCES:
            setattr(cls, "get_{}".format(resource), cls._partial('get_resource', tp_name))
            setattr(cls, "update_{}".format(resource), cls._partial('update_resource', tp_name))

    def _add_collection_methods(cls):
        for collection, tp_name in cls.COLLECTIONS:
            setattr(cls, "get_{}".format(collection), cls._partial('get_collection', tp_name))

    def _partial(cls, method_name, resource_name):
        def func(self, *args, **kwargs):
            return getattr(cls, method_name)(self, resource_name, *args, **kwargs)

        return func


@add_metaclass(MetaAPI)
class TargetProcessAPIClient(object):
    """Client that allows to retrieve and update data using TargetProcess API
    """
    ITEMS_PER_PAGE = 20
    MAX_RETRIES = 1

    def __init__(self, api_url, user, password):
        self.auth = HTTPBasicAuth(user, password)
        self.api_url = api_url
        self.query = {'format': 'json'}

    def get_resource(self, collection, entity_id, paginated=True, **query):
        url = self._build_resource_url(collection, entity_id, **query)
        if paginated:
            return self._get_paginated(url)
        else:
            return self._get(url)

    def get_collection(self, collection, paginated=True, **query):
        if not query.get('take'):
            query.update({'take': self.ITEMS_PER_PAGE})

        url = self._build_resource_url(collection, **query)
        if paginated:
            return self._get_paginated(url)
        else:
            return self._get(url)

    def update_resource(self, collection, entity_id, data):
        url = self._build_resource_url(collection, entity_id, include="[Id]")
        return self._post(url, data)

    def _build_resource_url(self, collection, entity_id=None, **query):
        if entity_id:
            path = '{0}/{1}'.format(collection, entity_id)
        else:
            path = '{}/'.format(collection)

        return self._build_url(path=path, **query)

    def _build_url(self, url=None, path=None, **query_params):
        if not url:
            url = self.api_url

        scheme, host, orig_path, params, orig_query, fragments = urlparse(url)

        query = self._build_query_params(orig_query, **query_params)
        path = "{0}/{1}".format(orig_path.rstrip('/'), path) if path else orig_path

        return urlunparse((scheme, host, path, params, query, fragments))

    def _build_query_params(self, orig_query, **additional_params):
        """
        Takes original query as a string (i.e `take=100&skip=25`)
        applies default query params from `self.query`
        adds and/or overrides from `**additional_params`
        encodes back to the url string
        """
        query = OrderedDict()
        query.update(parse_qsl(orig_query))
        query.update(sorted(additional_params.items()))
        query.update(sorted(self.query.items()))
        return urlencode(query)

    def _get(self, url):
        return self._do_request(method='get', url=url)

    def _get_paginated(self, url):
        response = self._get(url)
        if 'Items' in response:
            self._handle_pagination(response)
            for item in response.get('Items', []):
                self._handle_nested_pagination(item)
        else:
            self._handle_nested_pagination(response)
        return response

    def _handle_nested_pagination(self, item):
        nested_items = [(k, v) for k, v in item.items() if isinstance(v, dict)]
        for field, value in nested_items:
            self._handle_pagination(value)

    def _handle_pagination(self, data):
        if 'Next' in data:
            next_page = self._build_url(url=data['Next'])
            next_page_result = self._get(next_page)
            data['Items'] += next_page_result['Items']

    def _post(self, url, data):
        headers = {'Content-type': 'application/json'}
        json_data = json.dumps(data)
        return self._do_request(method='post', url=url, data=json_data, headers=headers)

    def _do_request(self, method, url, **kwargs):
        attempt = kwargs.pop('attempt', 0)
        try:
            request_method = requests.__getattribute__(method)
            response = request_method(url=url, auth=self.auth, **kwargs)
            if response.status_code not in (200, 201):
                raise BadResponseError(response=response)

            return response.json()

        except (ConnectionError, ReadTimeout) as e:
            if attempt < self.MAX_RETRIES:
                attempt += 1
                return self._do_request(url=url, method=method, attempt=attempt, **kwargs)

            logger.exception(e)
            return {}
